packages/browser/core/src/core/auth/internal/StitchAuthImpl.ts

```diff
@@ -140,12 +232,187 @@ export default class StitchAuthImpl extends CoreStitchAuth<StitchUser>
+  for (let i = 0; i < 64; ++i) {
+    state += alphaNumericCharacters.charAt(Math.floor(Math.random() * alphaNumericCharacters.length));
+  }
+  return state;
+}
+
+function pageRootUrl(): string {
+  return [
+    window.location.protocol,
+    "//",
+    window.location.host,
+    window.location.pathname
+  ].join("");
+}
+
+function unmarshallUserAuth(data): AuthInfo {
+  const parts = data.split("$");
+  if (parts.length !== 4) {
+    throw new StitchRedirectError(
+      "invalid user auth data provided while " +
+      "marshalling user authentication data: " + data
+    );
+  }
+
+  const [accessToken, refreshToken, userId, deviceId] = parts;
+  return new AuthInfo(userId, deviceId, accessToken, refreshToken);
+}
+
+class ParsedRedirectFragment {
+  public stateValid: boolean = false;
+  public authInfo?: AuthInfo;
+  public lastError?: string;
+  public clientAppIdValid: boolean = false;
+
+  get isValid(): boolean {
+    return this.stateValid && this.clientAppIdValid;
+  }
+}
+
+function parseRedirectFragment(fragment, ourState, ourClientAppId): ParsedRedirectFragment {
+  // After being redirected from oauth, the URL will look like:
+  // https://todo.examples.stitch.mongodb.com/#_stitch_state=...&_stitch_ua=...
+  // This function parses out stitch-specific tokens from the fragment and
+  // builds an object describing the result.
+  const vars = fragment.split("&");
+  const result: ParsedRedirectFragment = new ParsedRedirectFragment();
+  vars.forEach((kvp) => {
+  // for (let i = 0; i < vars.length; ++i) {
+    const pairParts = kvp.split("=");
+    const pairKey = decodeURIComponent(pairParts[0]);
+
+    switch (pairKey) {
+      case RedirectFragmentFields.StitchError:
+        result.lastError = decodeURIComponent(pairParts[1]);
+        break;
+      case RedirectFragmentFields.UserAuth:
+        try {
+          result.authInfo = unmarshallUserAuth(
+            decodeURIComponent(pairParts[1])
+          );
+        } catch (e) {
+          result.lastError = e;
+        }
+        break;
+      case RedirectFragmentFields.StitchLink:
+        break;
+      case RedirectFragmentFields.State:
+        const theirState = decodeURIComponent(pairParts[1]);
+
+        if (ourState === theirState) {
+          result.stateValid = true;
+        }
+        break;
+      case RedirectFragmentFields.ClientAppId:
+        const clientAppId = decodeURIComponent(pairParts[1]);
+        if (ourClientAppId === clientAppId) {
+          result.clientAppIdValid = true;
+        }
+        break;
+      default:
+        break;
+    }
+  });
+
+  return result;
 }
 
 function isAuthProviderClientFactory<ClientT>(
   factory:
     | AuthProviderClientFactory<ClientT>
     | NamedAuthProviderClientFactory<ClientT>
 ): factory is AuthProviderClientFactory<ClientT> {
-  return (<AuthProviderClientFactory<ClientT>>factory).getClient !== undefined;
+  return (factory as AuthProviderClientFactory<ClientT>).getClient !== undefined;
```

Comment:
did this need to change?

Reply:
my linker complains that the bracket format is deprecated for casting